src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java

```diff
@@ -32,76 +33,88 @@
-    }
-
-    @Override
-    public synchronized LogEntry read(final long position, final DirectFactory directFactory) {
-        return clone(read(position));
+    public synchronized int readTerm(final long index) {
+        return read(index).logKey().term();
     }
 
     @Override
-    public void readTo(final long position, final LogEntry logEntry) {
-        logEntry.copyFrom(read(position));
+    public synchronized void readKeyTo(final long index, final LogKey logKey) {
+        final LogKey readKey = read(index).logKey();
+        logKey.term(readKey.term()).index(readKey.index());
     }
 
     @Override
-    public void readKeyTo(final long position, final LogKey logKey) {
-        final LogKey readKey = read(position).logKey();
-        logKey.term(readKey.term()).index(readKey.index());
+    public synchronized void readTo(final long index, final LogEntry target) {
+        final LogEntry source = read(index);
+        target.writeBufferOrNull()
+                .putBytes(0, source.readBufferOrNull(), source.offset(), source.byteLength());
     }
 
-    private LogEntry read(final long position) {
-        if (position < entries.size()) {
-            return entries.get((int)position);
+    private LogEntry read(final long index) {
+        if (index < entries.size()) {
+            return entries.get((int)index);
         }
-        throw new IllegalArgumentException("Position " + position + " is out of bounds for size=" + entries.size());
+        throw new IllegalArgumentException("Index " + index + " is not in [0, " + (entries.size() - 1) + "]");
     }
 
     @Override
-    public synchronized void append(final LogEntry logEntry) {
-        entries.add(clone(logEntry));
+    public void append(final int term, final Command command) {
+        final LogEntry logEntry = directFactory.logEntry();
+        logEntry.logKey().term(term);
+        logEntry.command()
+                .writeBufferOrNull()
+                .putBytes(0, command.readBufferOrNull(), command.offset(), command.byteLength());
+        synchronized (this) {
+            final long index = entries.size();
+            logEntry.logKey().index(index);//TODO this is redundant, do we need this?
+            entries.add(logEntry);
+            maxCommandIndexBySourceId.put(command.commandKey().sourceId(), command.commandKey().commandIndex());
+        }
     }
 
 
     @Override
     public synchronized void truncateIncluding(final long index) {
-        if (index >= entries.size()) {
-            throw new IllegalArgumentException("Truncate index " + index + " must be less than the size " + entries.size());
-        }
-        for (long idx = lastIndex(); idx >= index; idx--) {
-            entries.remove((int)idx);
+        for (int idx = entries.size() - 1; idx >= index; idx--) {
+            final LogEntry removed = entries.remove(idx);
+            maxCommandIndexBySourceId.remove(removed.command().commandKey().sourceId());
```

Comment:
it does not look right to me. If you truncate the log there should be other commands from the SourceId left in the log, however the SourceId keyed entry would be removed.